packages/mendel-pipeline/src/pipeline.js

```diff
@@ -1,61 +1,96 @@
-    const ist = new IST({registry, transformer}, options);
-    const depsResolver = new DepResolver({registry, transformer}, options);
-
-    const steps = [watcher, reader, ist, depsResolver];
-    for (let i = 0; i < steps.length - 1; i++) {
-        const curStep = steps[i];
-        const nextStep = steps[i + 1];
-        curStep.on('done', function({entryId}) {
-            const entry = registry.getEntry(entryId);
-            entry.incrementStep();
-            nextStep.perform.apply(nextStep, [entry].concat(Array.prototype.slice.call(arguments, 1)));
-        });
+module.exports = class MendelPipeline extends EventEmitter {
+    constructor({defaultSteps, state, options}) {
+        super();
+        this._debug = _debug(`mendel:pipeline:${options.environment}`);
+
+        // Steps init
+        const ist = new IST(state, options);
+        const depsResolver = new DepResolver(state, options);
+
+        // Steps order
+        const steps = [
+            defaultSteps.watcher,
+            defaultSteps.reader,
+            ist,
+            depsResolver,
+        ];
+
+        // Async step chain
+        for (let i = 0; i < steps.length - 1; i++) {
+            const curStep = steps[i];
+            const nextStep = steps[i + 1];
+            curStep.on('done', function({entryId}) {
+                const rest = Array.prototype.slice.call(arguments, 1);
+                const entry = state.registry.getEntry(entryId);
+                entry.incrementStep();
+                nextStep.perform.apply(nextStep, [entry].concat(rest));
+            });
+        }
+
+        this.state = state;
+        this.steps = steps;
+        this.options = options;
     }
 
-    if (options.watch !== true) {
-        let totalEntries = 0;
-        let doneDeps = 0;
-
-        watcher.on('done', () => totalEntries++);
-        depsResolver.on('done', () => {
-            doneDeps++;
-            if (totalEntries === doneDeps) {
-                debug(`${totalEntries} entries were processed.`);
-                debug(
-                    Array.from(registry._mendelCache._store.values())
-                    .map(({id}) => id)
-                    .join('\n')
-                );
-                process.exit(0);
-            }
-        });
+    watch() {
+        this._debug('working');
+
+        let startedEntries = 0;
+        let doneEntries = 0;
+
+        this.steps[0]
+            .on('done', () => startedEntries++);
+
+        this.steps[this.steps.length-1]
+            .on('done', () => { doneEntries++;
+
+                if (startedEntries === doneEntries) {
+
+                    const total = this.state.registry.size();
+                    this._debug(`${doneEntries} entries were processed.`);
+                    this._debug(`${total} entries in registry.`);
+
+                    startedEntries = 0;
+                    doneEntries = 0;
+
+                    this.emit('idle', doneEntries);
+                }
+            });
```

Comment:
I think this is a healthy and stable lifecycle to debug on the daemon. I think we can keep this even if we decide later the daemon is not an emitter, or that the `idle` event should not exist.